added append-index-only-with-ingest-pipeline challenge to http_logs

jakelandis opened the pull request:
This change introduces a new challenge `append-index-only-with-ingest-pipeline` and corresponding operation `index-append-with-ingest-pipeline` and an additional corpora that mirrors the existing corpora. This is intended to help test the performance of the ingest node. This challenge requires `--elasticsearch-plugins="ingest-geoip" ` 

* `append-index-only-with-ingest-pipeline` and `index-append-with-ingest-pipeline` - runs the documents through an ingest node pipeline with the `grok` and `geoip` processors. 
* corpora `http_logs_unparsed` - identical data set to http_logs, but pieced back together to allow grok to break it into it's parts, specifically reshaped the data here: https://github.com/jakelandis/sample-data/tree/master/rally_http_logs, and uploaded to same location as http_logs, but with .unparsed. in the name.  The source documents for unparsed look like: 
```
{"message" : "30.87.8.0 - - [1998-05-24T15:00:01-05:00] \"GET /images/info.gif HTTP/1.0\" 200 1251"}
{"message" : "28.87.8.0 - - [1998-05-24T15:00:01-05:00] \"GET /french/images/hm_official.gif HTTP/1.1\" 200 972"}
{"message" : "17.87.8.0 - - [1998-05-24T15:00:01-05:00] \"GET /french/hosts/cfo/images/cfo/cfophot3.jpg HTTP/1.0\" 200 6695"}
{"message" : "72.236.7.0 - - [1998-05-24T15:00:01-05:00] \"GET /images/s102380.gif HTTP/1.1\" 200 207"}
{"message" : "82.36.0.0 - - [1998-05-24T15:00:02-05:00] \"GET /images/nav_bg_bottom.jpg HTTP/1.0\" 200 8389"}
```

To test locally from source
```
./gradlew -p plugins/ingest-geoip run
esrally --track-path=~/workspace/rally-tracks/http_logs --target-hosts=localhost:9200 --pipeline=benchmark-only --challenge=append-index-only-with-ingest-pipeline  --elasticsearch-plugins="ingest-geoip" --test-mode
```
To test this PR:

in rally.ini - add `jake.url = https://github.com/jakelandis/rally-tracks` under `[tracks]`
```
esrally --track=http_logs --track-repository=jake --challenge=append-index-only-with-ingest-pipeline  --elasticsearch-plugins="ingest-geoip" --test-mode
```

jakelandis commented:
@danielmitterdorfer @dliappis - I have made the changes as suggested in person (and a couple others) and have run the full tests though on my desktop with success. 

To test here are the commands (assuming you are pointed to my repo):
```
esrally --track=http_logs --track-repository=jake  --challenge=append-index-only-with-ingest-pipeline --track-params="ingest_pipeline:'baseline'"
esrally --track=http_logs --track-repository=jake  --challenge=append-index-only-with-ingest-pipeline --track-params="ingest_pipeline:'grok'"
esrally --track=http_logs --track-repository=jake  --challenge=append-index-only-with-ingest-pipeline --track-params="ingest_pipeline:'geoip'"  --elasticsearch-plugins="ingest-geoip"
```

* expose ingest_pipeline as track parameter
* decouples the performance metrics of grok and geoip by
exposing the processor in different pipeline.
* introduces of a baseline ingest pipeline that does nothing
so that the overhead of the pipeline may be measured.
* fixes the date mapping to correctly account for
epoch_seconds.
* introduces the script used to create the unparsed data
set.

danielmitterdorfer commented:
Thanks for the update @jakelandis. I'll test this soon.

Note to self: This PR will close https://github.com/elastic/rally-tracks/issues/33.

jakelandis commented on http_logs/operations/default.json:
```diff
@@ -2,7 +2,32 @@
+      "bulk-size": {{bulk_size | default(5000)}},
+      "ingest-percentage": {{ingest_percentage | default(100)}},
+      "pipeline": "http-log-geoip-pipeline",
+      "corpora": "http_logs"
```
this needs the pre-parsed corpus, else we need to parse data with with grok to get the ip as it's own field.

danielmitterdorfer replied:
Good point. I missed that somehow. You're right.

jakelandis commented on http_logs/operations/default.json:
```diff
@@ -2,7 +2,32 @@
+      "bulk-size": {{bulk_size | default(5000)}},
+      "ingest-percentage": {{ingest_percentage | default(100)}},
+      "pipeline": "http-log-baseline-pipeline",
+      "corpora": "http_logs"
```
it is using the `http_logs` ?

_assuming you meant `http_logs_unparsed` ...
It uses the pre-parsed version so not occur the over head of parsing, but still occur the overhead of indexing all of the fields. If I were to use the `http_logs_unparsed` it would end up indexing only the message field which is actually index:false.  The net difference between the baseline with pre-parsed vs. grok with un-parsed should be the overhead of grok plus the (un-avoidable) overhead of the index:false message field.

danielmitterdorfer replied:
Maybe I should have not stated the obvious... . What I wanted to convey here is that my initial thought was that you should use `http_logs_unparsed`  but then I realized that `http_logs` is indeed correct... . Sorry for the confusion. :)

jakelandis commented:
> note that I've already corrected the corpus for my tests so I ran it with the http_logs_unparsed corpus

thanks!  did you upload the fixed .bz2 ?  if not I can hunt em down and fix it.  

> One thing I am a bit concerned about is that Rally will now download both corpora even if it needs only one of them.

I noticed that too, let me know if you want any change to this PR to help with that concern.

danielmitterdorfer commented:
>> note that I've already corrected the corpus for my tests so I ran it with the http_logs_unparsed corpus

> thanks! did you upload the fixed .bz2 ? if not I can hunt em down and fix it.

By "corrected" I meant that I changed the referenced corpus from `http_logs` to `http_logs_unparsed`  for the geoip pipeline (which was wrong anyway as you noted so disregard my comment). I did not check in more detail what documents were problematic for the grok pipeline.

>> One thing I am a bit concerned about is that Rally will now download both corpora even if it needs only one of them.

> I noticed that too, let me know if you want any change to this PR to help with that concern.

It would be great if you could implement the workaround that I've mentioned. Basically you need to wrap the corpus definition in a Jinja `if` - `else`. I also think that for the (temporary) workaround it is ok if you use the existing track parameter `ingest_pipeline`. Something along the lines of this should do the trick (untested):

```
{%- if ingest_pipeline is defined and ingest_pipeline == "grok" %}
    {
      "name": "http_logs_unparsed",
      ...
    }
{%- else %}
    {
      "name": "http_logs",
      ...
    }
{%- endif %}
```

Note that this is only intended as a temporary workaround and I have raised https://github.com/elastic/rally/issues/553 to fix this properly in Rally itself.

I think the remaining points for this PR are:

* Fix the unparsed document files so the error rate is zero.
* Add this workaround

Thanks so much for doing this @jakelandis.

jakelandis commented:
@danielmitterdorfer - changes made as requested. I found (via grep) and fixed (manually) ~15 entries that had a rouge `\` which made the Json source invalid. I added a TODO in the unparser.rb to validate for proper JSON in case we ever need to run it again. 

Grok now runs without any errors and the corpora is only downloaded as needed.

jakelandis commented:
@danielmitterdorfer @dliappis thanks for the feedback and prompt reviews. merging ...
